doc/oml2-proxy-server.1.txt
oml2-proxy-server(1)
====================

NAME
----
oml2-proxy-server - Buffer OML client measurements for later replay
to an upstream server.

SYNOPSIS
--------
[verse]
*oml2-proxy-server* [-l port | --listen=port] [-r file | --resultfile=file]
      [-s size| --size=size] [-a addr | --dstaddress=addr] [-p port | --dstport=port]
	  [-d level | --debug-level=level] [--logfile=file] [-v | --version]
	  [-? | --help]

DESCRIPTION
-----------

*oml2-proxy-server* provides a buffered proxy for the linkoml:oml2-server[1]
for use in situations where either the normal *oml2-server* is not
reachable, or where using a reachable *oml2-server* to store
measurements as they are made would disturb the phenomena being
measured.

The *oml2-proxy-server* listens for OML client connections on the port
given with the -l option, and buffers the subsequent measurement
traffic in memory. When it receives an OML-RESUME command (see below),
*oml2-proxy-server* connects to the upstream *oml2-server* daemon
(specified with the -a and -p options) and then sends all buffered
measurements to the upstream server.

The buffering scheme is very simple: the client measurement streams
are stored in the buffer without translation, and then simply replayed
to the upstream server when the proxy is resumed.  This allows the
*oml2-proxy-server* implementation to be very simple.

The *oml2-proxy-server* listens for the OMLPROXY-PAUSE and
OMLPROXY-RESUME commands on its stdin file descriptor.  The operator
must therefore leave the process attached to an open terminal or pipe
whilst it is running in order to control it.  The commands must be
typed in upper case as listed here, followed by a carriage return
(ENTER key).

The OMLPROXY-PAUSE command pauses the transmission of measurements to
the upstream server.  When *oml2-proxy-server* starts, it is in this
state.

The OMLPROXY-RESUME command resumes transmission of measurements to
the upstream server.

A typical usage scenario is to start the *oml2-proxy-server* and leave
it in the paused state, then conduct measurements, and when
measurement activities have finished, issue the OMLPROXY-RESUME
command to order the proxy server to send measurements to the upstream
OML server.

As well as storing the measurement streams in memory, the
*oml2-proxy-server* also saves them to a file on disk, whose name and
location can be specified with the '-r' option.

OPTIONS
-------
-l port::
--listen=port::
	Listen for connections from OML clients on port (default 3003).

-d level::
--debug-level=level::
	Set the verbosity of the log output to level.  The level should be
	an integer from 1 to 4 (1=ERROR, 2=WARNING, 3=INFO, 4=DEBUG).  The
	default log verbosity is 3, INFO.

--logfile=file::
	Write log messages to the given file.  The default file is
	oml_proxy_server.log.

-r file::
--resultfile=file::
	Save results into the given disk file.  The default result file
	name is oml_result_proxy.log.  If the result file already exists
	then a new one will be created with a unique name based on file.

-s bytes::
--size=bytes::
	Buffer page size in bytes.

-p port::
--dstport=port::
	Upstream server port number (default port is 3003).

-a address::
--dstaddress=address::
	Upstream server address (default is localhost).

-v::
--version::
	Print the version number of *oml2-proxy-server*.

-?::
--help::
	Print a summary of available options.

--usage::
	Display a brief usage message.


BUGS
----
The *oml2-proxy-server* is still experimental software and under
active development.  It may contain bugs, and its interfaces may
change without warning from version to version, even within the 2.x
major release series.  We will strive to keep the interface backwards
compatible for patch-level releases within a minor release series
(i.e. version 2.3.5 should be backwards compatible with 2.3.4), but we
may change the interface between minor release series (e.g. version
2.4.0 might not be backwards compatible with 2.3.5).

include::bugs.txt[]

SEE ALSO
--------
Manual Pages
~~~~~~~~~~~~
linkoml:oml2-server[1], linkoml:liboml2[1], linkoml:liboml2[3]

include::manual.txt[]

// vim: ft=asciidoc:tw=72


ChangeLog
2012-11-23  Olivier Mehani  <olivier.mehani@nicta.com.au>
	* OML: Version 2.9.0rc
	* liboml2: Bump to library version 9.0.0 (1.0.0 for OComm)
		Heavy work on memory leak issues (#647, #516)
		Changes in the OmlValueU API
			Old applications should be ported by using omlc_zero(3)
			or omlc_zero_array(3) after declaring OmlValueUs, and
			properly omlc_reset_string(3) and omlc_reset_blob(3)
			when needed. All the API is documented in OmlValueU(3),
			but should not be used directly with new application.
			see OmlValue(3) for more details on porting pre-2.9
			instrumentations.
		Instrumentation workflow MUST now use oml2-scaffold(1) and
			oml_inject_MPNAME(3).
		Old binaries will still work with the old library which can be
			installed in parallel
		Re-insteate OML's functionality of being able to
			delegate logging to the application (#832)
		Throttle logging when too many messages (#827)
		More informative and regular log messages (#222)
		Plenty of bugfixes
	* UI changes: Introduce --oml-domain and --oml-collect instead of
		(resp.) --oml-exp-id and --oml-server, which are still valid
		(#860, #726)
	* Documentation: #891, #771
			 Update manpages
			 Document some parts of the code with Doxygen
			 A README file!
	* Bindings: New OML4Py Python module contributed by Fraida Fund from
		NYU-Poly (#595)
		    OML4R now supports reconnection
	* Supported architectures: Basic Android port thanks to François Hoguet
		from UTC (#512)
	* Supported backends: PostgreSQL is officially supported (#874, #859,
		#825, #824, #755)

2012-06-25  Olivier Mehani <olivier.mehani@nicta.com.au>
	* OML: Version 2.8.0
	* Bugfixes: Builds on Mac OS X without special arguments (#749)
	* liboml2: Reconnect to server in case of disconnection (#757)
	* oml2-server: Fix issue with empty string last in text mode (#610)
		       Introduce event hook mechanism to react to event within
		       the server (e.g., push a measurement database to iRODS
		       when the last client has disconnected)
	* oml2-scaffold: Load the right appfile (#750)
	* OML4R: OML4R Gem pushed to RubyGems
		 Examples updated
		 Support for sending measurement to different streams
		 Make parameters constistent with liboml and command line (#751)

2012-04-24  Olivier Mehani <olivier.mehani@nicta.com.au>
	* OML: Version 2.7.0-rc1
	* manpages: Fix manpage wrt --oml-list-filters
	            Fix server's data-dir option prototype
	* liboml2: Output to local file by default if nothing specified
	           Use simplified logging style by default
	           Unify some debug messages between bin and text
	           Implement the flush:// scheme for the file URI
	           Avoid compiler warning with omlc_set_const_string()
	* oml2-scaffold: Use OML_INT32_VALUE by default
		         Exit with an error on unknown types
		         Generate per-MP injector functions
		         Generate .rb app files
		         Fix popt.h generation
		         Fix Makefile generation
	                 Handle both :type and "type"
			 Brand new example oml2-scaffold-generated generator

2011-06-19  Jolyon White  <jolyon.white@nicta.com.au>

	* OML: New version 2.6.0

	* lib/client/init.c (create_writer): The output encoding is now
	independent of the output transport.  Text encoding can be sent to
	network, and binary encoding can be sent to local file, as well as
	the existing options (text->file, binary->network).  Encoding is
	selectable using --oml-text and --oml-binary.

	* lib/client/buffered_writer.c: Add new output serialization and
	buffering scheme to the client library to decouple backend
	processing from application-thread sample injects.  This should
	make measurements have less impact on the main thread.  New client
	lib command line option --oml-bufsize controls the amount of
	buffer space available in the buffer that sits between the inject
	thread and the output stage thread.

	* server/main.c (drop_privileges): oml2-server now drops
	privileges if the --user and/or --group command line options.
	Privileges are dropped as soon as possible, namely after the
	socket has been set up (in case the admin wants to run it on a
	privileged port) and after the log file has been opened.
	(setup_backend_sqlite): oml2-server now writes databases to a
	subdirectory of $(localstatedir), which can be configured at
	compile time using the --localstatedir=DIR option of configure.
	If the server can't write to that directory, it will abort with an
	error message in the log.  The default is overridden by the
	OML_SQLITE_DIR environment variable, which is overridden by the
	--data-dir option (now available as -D as well).

	* server/psql_adapter.c (psql_create_database): PostgreSQL support
	is not complete and considered to be supported.  oml2-server now
	supports --pg-connect and --pg-user to help configure the
	connection to the PostgreSQL server.

	* server/main.c (setup_logging): Make oml2-server and
	oml2-proxy-server log to stderr when stderr is attached to a tty
	if --logfile is not used.  If --logfile is not used and stderr is
	NOT attached to a tty, the servers log to their default log file
	names instead.

	* lib/client/parse_config.c: reskin the XML configuration file
	format to make the element and attribute names more meaningful,
	and more easily remembered.

2011-03-16  Jolyon White  <jolyon.white@nicta.com.au>

	* OML: New version 2.5.2

	* oml2-server: All strings representing table and column names
	are now enclosed in double quotes to allow SQL keywords to be
	used as column names.  Note that the strings are not yet escaped,
	so including a double quote in the name can be used to do SQL
	injection.

	* oml2-server: Fix a bug in copying string values from the text
	protocol parser that resulted in incorrectly terminated
	strings. This would sometimes add extra garbage to the end of
	strings stored in the database.

2011-01-29  Jolyon White  <jolyon.white@nicta.com.au>

	* OML: New version 2.5.1


proxy_server/oml2-proxy-server.c
/*
 * Copyright 2007-2012 National ICT Australia (NICTA), Australia
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>
#include <signal.h>
#include <popt.h>

#include "ocomm/o_log.h"
#include "ocomm/o_socket.h"
#include "ocomm/o_eventloop.h"
#include "mstring.h"
#include "session.h"
#include "proxy_client.h"

#define V_STRING  "OML2 Proxy Server V%s\n"
#define COPYRIGHT "Copyright 2007-2012 NICTA\n"

#define DEF_PORT 3003
#define DEF_PORT_STR "3003"
#define DEFAULT_LOG_FILE "oml_proxy_server.log"
#define DEFAULT_RESULT_FILE "oml_result_proxy.res"
#define DEF_PAGE_SIZE 1024
#define DEFAULT_SERVER_ADDRESS "localhost"

static int listen_port = DEF_PORT;

static int log_level = O_LOG_INFO;
static char* logfile_name = NULL;
static char resultfile_name[128] = DEFAULT_RESULT_FILE;
static int page_size = DEF_PAGE_SIZE;
static int downstream_port = DEF_PORT;
static char* downstream_address = DEFAULT_SERVER_ADDRESS;
int sigpipe_flag = 0; // Set to 'true' by signal handler.

Session* session = NULL;
extern void *client_send_thread (void* handle);


struct poptOption options[] = {
  POPT_AUTOHELP
  { "listen",      'l',  POPT_ARG_INT,    &listen_port,     0,   "Port to listen for TCP based clients", DEF_PORT_STR},
  { "debug-level", 'd',  POPT_ARG_INT,    &log_level,       0,   "Debug level - error:1 .. debug:4",     NULL},
  { "logfile",     '\0', POPT_ARG_STRING, &logfile_name,    0,   "File to log to",                       DEFAULT_LOG_FILE },
  { "version",     'v',  POPT_ARG_NONE,   NULL,             'v', "Print version information and exit",   NULL},
  { "resultfile",  'r',  POPT_ARG_STRING, &resultfile_name, 0,   "File name for storing received data",  DEFAULT_RESULT_FILE},
  { "size",        's',  POPT_ARG_INT,    &page_size,       0,   "Page size for buffering measurements", NULL},
  { "dstport",     'p',  POPT_ARG_INT,    &downstream_port, 0,   "Downstream OML server port",       NULL},
  { "dstaddress",  'a',  POPT_ARG_STRING, &downstream_address,  0,   "Downstream OML server address",    DEFAULT_SERVER_ADDRESS },
  { NULL,          0,    0,               NULL,             0,   NULL,                                   NULL }
};

void
proxy_message_loop (const char *client_id, Client *client, void *buf, size_t size);

/**
 * \brief function called when the socket receive some data
 * \param source the socket event
 * \param handle the cleint handler
 * \param buf data received from the socket
 * \param bufsize the size of the data set from the socket
 */
void
client_callback(SockEvtSource *source, void *handle, void *buf, int buf_size)
{
  (void)source;
  Client* self = (Client*)handle;

  //  logdebug ("'%s': received %d octets of data\n", source->name, buf_size);
  //  logdebug ("'%s': %s\n", source->name, to_octets (buf, buf_size));

  proxy_message_loop (source->name, self, buf, buf_size);

  mbuf_repack_message (self->mbuf);

  if (self->state == C_PROTOCOL_ERROR) {
    socket_close (self->recv_socket);
    logerror("'%s': protocol error, proxy server will disconnect upstream client\n", source->name);
    eventloop_socket_remove (self->recv_event);  // Note:  this free()'s source!
    self->recv_event = NULL;
  }

  fwrite (buf, sizeof (char), buf_size, self->file);

  pthread_mutex_lock (&self->mutex);
  pthread_cond_signal (&self->condvar);
  pthread_mutex_unlock (&self->mutex);
}

/**
 * \brief Call back function when the status of the socket change
 * \param source the socket event
 * \param status the status of the socket
 * \param error the value of the error if there is
 * \param handle the Client handler structure
 */
void
status_callback(SockEvtSource *source, SocketStatus status, int error, void *handle)
{
  (void)error;
  switch (status) {
    case SOCKET_CONN_CLOSED: {
      Client* self = (Client*)handle;
      if (self->recv_event != NULL) {
        socket_close (source->socket);
        logdebug("socket '%s' closed\n", source->name);
        eventloop_socket_remove (source); // Note:  this free()'s source!
      }

      /* Signal the sender thread for this client that the client disconnected */
      pthread_mutex_lock (&self->mutex);
      self->state = C_DISCONNECTED;
      pthread_cond_signal (&self->condvar);
      pthread_mutex_unlock (&self->mutex);
      break;
    default:
      break;
    }
  }
}

/**
 * \brief Called when a node connects via TCP
 * \param
 * \return
 */
void
on_connect (Socket* client_sock, void* handle)
{
  (void)handle;  // This parameter is unused
  MString *mstr = mstring_create ();

  mstring_sprintf (mstr,"%s.%d", resultfile_name, session->client_count);
  logdebug("New client (index %d) connected\n", session->client_count);
  session->client_count++;

  Client* client = client_new(client_sock, page_size, mstring_buf (mstr),
                              downstream_port, downstream_address);

  mstring_delete (mstr);

  session_add_client (session, client);
  client->session = session;

  client->recv_event = eventloop_on_read_in_channel(client_sock, client_callback,
                                                    status_callback, (void*)client);

  pthread_create (&client->thread, NULL, client_send_thread, (void*)client);
}

/*
 *  A hack to work around Mac OSX's broken implementation of poll(2),
 *  which does not allow polling stdin.  This hack dup()'s stdin to
 *  another file descriptor and creates a pipe(2) which it then
 *  dup2()'s into the normal stdin file descriptor.  If this succeeds,
 *  it then spawns a thread whose only function is to read(2) from the
 *  original stdin and write into the pipe.  The read end of the pipe,
 *  which is now file descriptor 0, is then useable in a call to
 *  poll(2).
 */
void*
prepare_stdin (void *handle)
{
  (void)handle;
#ifdef __APPLE__
  static int thread_started = 0;
  static int stdin_dup;
  static int stdin_pipe[2];
  static pthread_t self;
  if (!thread_started) {
    int result;
    result = pipe (stdin_pipe);
    if (result) {
      logerror ("Could not create pipe for stdin duplication: %s\n", strerror(errno));
      exit(EXIT_FAILURE);
    }

    stdin_dup = dup(fileno(stdin));
    dup2(stdin_pipe[0], fileno(stdin));

    thread_started = 1;
    result = pthread_create (&self, NULL, prepare_stdin, NULL);
    if (result == -1) {
      logerror ("Error creating thread for reading stdin: %s\n", strerror (errno));
      exit (EXIT_FAILURE);
    }
  } else {
    char buf [512];
    while (1) {
      ssize_t count = read (stdin_dup, buf, 512);
      if (count < 0)
        logerror ("Error reading from stdin: %s\n", strerror(errno));
      if (count <= 0)
        return NULL;
      write (stdin_pipe[1], buf, count);
    }
  }
#endif
  return NULL;
}

void
stdin_handler(SockEvtSource* source, void* handle, void* buf, int buf_size)
{
  Session *proxy = (Session*)handle;
  char command[80];
  (void)source;
  strncpy (command, buf, 80);

  if (buf_size < 80 && command[buf_size-1] == '\n')
    command[buf_size-1] = '\0';

  printf ("Received command: %s\n", command);
  logdebug ("Received command: %s\n", command);

  if ((strcmp (command, "OMLPROXY-RESUME") == 0) ||
      (strcmp (command, "RESUME") == 0)) {
    proxy->state = ProxyState_SENDING;
  } else if (strcmp (command, "OMLPROXY-STOP") == 0 ||
             strcmp (command, "STOP") == 0) {
    proxy->state = ProxyState_STOPPED;
  } else if (strcmp (command, "OMLPROXY-PAUSE") == 0 ||
             strcmp (command, "PAUSE") == 0) {
    proxy->state = ProxyState_PAUSED;
  }

  /*
   * If we're in sending state, wake up the client sender threads
   * so that they will start sending to the downstream server.  We do
   * this even if the state was already ProxyState_SENDING because
   * some of the clients might have dropped back to idle due to
   * disconnection from the upstream server.
   */
  if (session->state == ProxyState_SENDING) {
    Client *current = session->clients;
    while (current) {
      pthread_mutex_lock (&current->mutex);
      pthread_cond_signal (&current->condvar);
      pthread_mutex_unlock (&current->mutex);
      current = current->next;
    }
  }
}

void
on_control_connect (Socket *sock, void *handle)
{
  (void)handle; // This parameter is unused
  logdebug ("New control connection\n");
  eventloop_on_read_in_channel (sock, stdin_handler, NULL, (void*)session);
}

void
sigpipe_handler(int signum)
{
  if (signum == SIGPIPE)
    sigpipe_flag = 1;
}

/**
 * @brief Set up the logging system.
 *
 * This function sets up the server logging system to log to file
 * logfile, with the given log verbosity level.  All messages with
 * severity less than or equal to level will be logged; all others
 * will be discarded (lower levels are more important).
 *
 * If logfile is not NULL then the named file is opened for logging.
 * If logfile is NULL and the application's stderr stream is not
 * attached to a tty, then the file DEF_LOG_FILE is opened for
 * logging; otherwise, if logfile is NULL and stderr is attached to a
 * tty then log messages will sent to stderr.
 *
 * @param logfile the file to open
 * @param level the severity level at which to log
 */
void
setup_logging (char *logfile, int level)
{
  if (!logfile) {
    if (isatty (fileno (stderr)))
      logfile = "-";
    else
      logfile = DEFAULT_LOG_FILE;
  }

  o_set_log_file(logfile);
  o_set_log_level(level);
  o_set_simplified_logging ();
}

int
main(int argc, const char *argv[])
{
  Socket* serverSock, *controlSock;
  int c;

  poptContext optCon = poptGetContext(NULL, argc, argv, options, 0);
  poptSetOtherOptionHelp(optCon, "configFile");

  while ((c = poptGetNextOpt(optCon)) >= 0) {
    switch (c) {
    case 'v':
      printf(V_STRING, VERSION);
      printf(COPYRIGHT);
      return 0;
    }
  }

  setup_logging (logfile_name, log_level);

  if (c < -1) {
    /* an error occurred during option processing */
    fprintf(stderr, "%s: %s\n",
      poptBadOption(optCon, POPT_BADOPTION_NOALIAS),
      poptStrerror(c));
    return -1;
  }

  loginfo (V_STRING, VERSION);
  loginfo (COPYRIGHT);

  struct sigaction new_action, old_action;
  new_action.sa_handler = sigpipe_handler;
  sigemptyset (&new_action.sa_mask);
  new_action.sa_flags = 0;
  sigaction (SIGPIPE, NULL, &old_action);
  if (old_action.sa_handler != SIG_IGN)
    sigaction (SIGPIPE, &new_action, NULL);

  prepare_stdin(NULL);

  eventloop_init();
  session = (Session*) malloc(sizeof(Session));
  memset(session, 0, sizeof(Session));

  session->state = ProxyState_PAUSED;

  serverSock = socket_server_new("proxy_server", listen_port, on_connect, NULL);
  controlSock = socket_server_new("proxy_server_control", listen_port + 1, on_control_connect, NULL);

  eventloop_on_stdin(stdin_handler, session);
  eventloop_run();

  socket_free(serverSock);
  socket_free(controlSock);

  return(0);
}

/*
 Local Variables:
 mode: C
 tab-width: 2
 indent-tabs-mode: nil
 End:
 vim: sw=2:sts=2:expandtab
*/
